### PR TITLE
Add support for Column(Order) in order by.

### PR DESCRIPTION
**Pull request by ConstantinOr**
### Issues

*This pull request fixes issue #1085.*

### Description

By default if Column(Order) present in model, "order by" created based in  alphabetical order.
This code corrected  this issue,  now "order by" will creating based on Column(Order) value.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

Bug 1085 was fixed, no additional work in necessary.

### CODE REVIEW

**Comment by ConstantinOr** [comment]
Please explain what wrong with checks. What need to do for successful.

**Comment by ConstantinOr** [comment]
@biaol-odata Hello. Can you help me to understand what wrong with my push request, why all checks was not pass?

**Comment by xuzhg** [comment]
@ConstantinOr Build system shows the error message as below:

##[error]src\System.Web.OData\OData\Builder\Conventions\Attributes\ColumnAttributeEdmPropertyConvention.cs(53,1): Error : SA1508: A closing curly bracket must not be preceded by a blank line

**Comment by xuzhg** [comment]
@ConstantinOr also please make sure all test passed, UT & E2E.

for example, the build system shows you should modify `System.Web.OData.PublicApi.PublicApiTest.TestPublicApi` to make it pass.

**Comment by robward-ms** [comment]
New build error:

2017-10-23T20:38:09.6716702Z   ODataOrderByTest\OrderByTest.cs(55,33): error CS1056: Unexpected character '$' [C:\BuildAgent\_work\155\s\test\E2ETest\WebStack.QA.Test.OData\WebStack.QA.Test.OData.csproj]
2017-10-23T20:38:09.6716702Z   ODataOrderByTest\OrderByTest.cs(70,63): error CS1056: Unexpected character '$' [C:\BuildAgent\_work\155\s\test\E2ETest\WebStack.QA.Test.OData\WebStack.QA.Test.OData.csproj]

**Comment by robward-ms** [comment]
You can use the build.cmd in the root directory to compile all code and run all tests.

File: test/E2ETest/WebStack.QA.Test.OData/ODataOrderByTest/OrderByTest.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                routePrefix: "odata",
+                model: OrderByEdmModel.GetModel());
+
+            configuration.EnsureInitialized();
+        }
+
+        [Fact]
+        public async Task TestOrderByResult()
+        {   // Arrange
+            string requestUri = $"{BaseAddress}/odata/Items";
```
Comment (robward-ms) [inline test/E2ETest/WebStack.QA.Test.OData/ODataOrderByTest/OrderByTest.cs]: Please replace this with a non-interpolated string. Same line 70.

**Comment by ConstantinOr** [comment]
@robward-ms Issue fixed but no luck with build.

**Comment by robward-ms** [comment]
@ConstantinOr - The compile portion of the build succeeded but there were lots of test failures. I verified locally (pulled your code, ran tests in VS) that there is not* a test problem with your change so I'm re-running a build.

When running build.cmd, I do see one code analysis error so I added a comment here for the fix needed to pass that. It seems we don't have this well integrated into our builds, only on the cmd line build.

Locally, I did see one error in your new E2E test, TestOrderByResult: 

Test Name:	WebStack.QA.Test.OData.ODataOrderByTest.ODataOrderByTest.TestOrderByResult
Test FullName:	WebStack.QA.Test.OData.ODataOrderByTest.ODataOrderByTest.TestOrderByResult
Test Source:	e:\OData\GitHub\robward-ms\WebApi\test\E2ETest\WebStack.QA.Test.OData\ODataOrderByTest\OrderByTest.cs : line 54
Test Outcome:	Failed
Test Duration:	0:00:00.863

Result Message:	System.Exception : Cannot setup host server. See Event log for details
Result StackTrace:	
at Nuwa.WebStack.Host.KatanaSelfHostElement.InitializeServer(RunFrame frame) in e:\OData\GitHub\robward-ms\WebApi\test\E2ETest\Nuwa.WebStack\Host\KatanaSelfHostElement.cs:line 67
   at Nuwa.Sdk.Elements.BaseHostElement.Initialize(RunFrame frame) in e:\OData\GitHub\robward-ms\WebApi\test\E2ETest\Nuwa.WebStack\Host\BaseHostElement.cs:line 23
   at Nuwa.Sdk.RunFrame.Initialize(Object testClass, NuwaTestCommand testCommand) in e:\OData\GitHub\robward-ms\WebApi\test\E2ETest\Nuwa\Sdk\RunFrame.cs:line 51
   at Nuwa.Sdk.NuwaTestCommand.Execute(Object testClass) in e:\OData\GitHub\robward-ms\WebApi\test\E2ETest\Nuwa\Sdk\NuwaTestCommand.cs:line 58

File: src/System.Web.OData/OData/Builder/Conventions/Attributes/ColumnAttributeEdmPropertyConvention.cs [hunk 2]
```diff
@@ -36,13 +38,23 @@ public ColumnAttributeEdmPropertyConvention()
...
+            var primitiveProperty = edmProperty as PrimitivePropertyConfiguration;           
             if (primitiveProperty == null)
             {
                 return; // ignore non-primitive property
             }
 
             var columnAttribute = attribute as ColumnAttribute;
+            if (columnAttribute != null && columnAttribute.Order > 0)
+            {
+                if (ODataQueryOptions.ColumnOrder == null)
```
Comment (robward-ms) [inline src/System.Web.OData/OData/Builder/Conventions/Attributes/ColumnAttributeEdmPropertyConvention.cs]: You can get rid of this.

File: src/System.Web.OData/OData/Query/ODataQueryOptions.cs [hunk 3]
```diff
@@ -87,6 +87,11 @@ public ODataQueryOptions(ODataQueryContext context, HttpRequestMessage request)
         }
 
         /// <summary>
+        /// Gets Column order.
+        /// </summary>
+        public static Dictionary<string, int> ColumnOrder { get; set; }
```
Comment (robward-ms) [inline src/System.Web.OData/OData/Query/ODataQueryOptions.cs]: Change to private set and initialize to new Dictionary<string, int>() in the constructor. The code analysis error says to make this read-only. You don't need to let another class set if equally to a new dictionary, only to populate it which can be achieved with a read-only property.

File: src/System.Web.OData/OData/Builder/Conventions/Attributes/ColumnAttributeEdmPropertyConvention.cs [hunk 4]
```diff
@@ -36,13 +38,23 @@ public ColumnAttributeEdmPropertyConvention()
                 return;
             }
 
-            var primitiveProperty = edmProperty as PrimitivePropertyConfiguration;
+            var primitiveProperty = edmProperty as PrimitivePropertyConfiguration;
```
Comment (robward-ms) [inline src/System.Web.OData/OData/Builder/Conventions/Attributes/ColumnAttributeEdmPropertyConvention.cs]: Can you also delete the trailing spaces here?

Comment (ConstantinOr) [reply src/System.Web.OData/OData/Builder/Conventions/Attributes/ColumnAttributeEdmPropertyConvention.cs]: All done.

**Comment by robward-ms** [comment]
@ConstantinOr - I've had a look at your change and I now know what's happening.

Once we got unblocked by build errors, there was fairly massive test failure. This was caused by ODataQueryOptions.ColumnOrder being set to null because it was only initialized in the case where a ColumnAttribute was present in ColumnAttributeEdmPropertyConvention.

With the recent change, where ColumnOrder  is initialized in the ODataQueryOptions constructor, we regularly get 2 test failures: your new E2E test and the public API unit test. The public API unit test. The public API unit test is easy to explain: you added the ColumnOrder property to a public class and added new API without updating the public API baseline. When you run this test, it will fail and explain how to fix it, i.e. copy the new public API output to the baseline file.

Your new test is failing because you now initialize ColumnOrder in the ODataQueryOptions constructor but ColumnOrder is static; the code in ColumnAttributeEdmPropertyConvention now throws an exception because ColumnOrder is null, which causes model generation to fail which occurs when the test setups the web config, which results in the test not starting.

For your development, you should be running the local build.cmd to ensure all code compiles and all tests pass. At a minimum, you should be running all tests in VS using our WebApiOData and WebApiOData.E2e solutions.

File: src/System.Web.OData/OData/Query/ODataQueryOptions.cs [hunk 5]
```diff
@@ -84,9 +84,16 @@ public ODataQueryOptions(ODataQueryContext context, HttpRequestMessage request)
...
 
             Validator = ODataQueryValidator.GetODataQueryValidator(context);
+
+            ColumnOrder = new Dictionary<string, int>();
         }
 
         /// <summary>
+        /// Gets Column order.
+        /// </summary>
+        public static Dictionary<string, int> ColumnOrder { get; private set; }
```
Comment (robward-ms) [inline src/System.Web.OData/OData/Query/ODataQueryOptions.cs]: Why is this static? Do you really want to share the same dictionary for all objects in the model?

File: src/System.Web.OData/Settings.StyleCop [hunk 6]
```diff
@@ -1,29 +1,15 @@
 <StyleCopSettings Version="105">
+  <GlobalSettings>
```
Comment (robward-ms) [inline src/System.Web.OData/Settings.StyleCop]: Why are you modifying the stylecop configuration?

File: test/E2ETest/WebStack.QA.Test.OData/ODataOrderByTest/OrderByEdmModel.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+            public IDbSet<Item> Items { get; set; }
+        }
+        
+
+        public static IEdmModel GetModel()
+        {
+            ODataConventionModelBuilder builder = new ODataConventionModelBuilder();
+            builder.EntitySet<Item>("Items");
+
```
Comment (robward-ms) [inline test/E2ETest/WebStack.QA.Test.OData/ODataOrderByTest/OrderByEdmModel.cs]: please add a second model with different ordering to ensure it works with more than one model.

File: src/System.Web.OData/Settings.StyleCop (hunk 6 above)
Comment (ConstantinOr) [reply src/System.Web.OData/Settings.StyleCop]: Sorry my mistake.

File: src/System.Web.OData/OData/Query/ODataQueryOptions.cs (hunk 5 above)
Comment (ConstantinOr) [reply src/System.Web.OData/OData/Query/ODataQueryOptions.cs]: I think this is more simple way to store column Order.

**Comment by ConstantinOr** [comment]
@robward-ms  All was done. Open Api fixed, was added new model and test.

File: src/System.Web.OData/Settings.StyleCop (hunk 6 above)
Comment (robward-ms) [reply src/System.Web.OData/Settings.StyleCop]: This file is still included in the commit, did you plan to revert it?

File: src/System.Web.OData/OData/Builder/Conventions/Attributes/ColumnAttributeEdmPropertyConvention.cs [hunk 8]
```diff
@@ -43,6 +45,11 @@ public ColumnAttributeEdmPropertyConvention()
             }
 
             var columnAttribute = attribute as ColumnAttribute;
+            if (columnAttribute != null && columnAttribute.Order > 0)
+            {
+                ODataQueryOptions.ColumnOrder.Add(edmProperty.Name, columnAttribute.Order);
```
Comment (robward-ms) [inline src/System.Web.OData/OData/Builder/Conventions/Attributes/ColumnAttributeEdmPropertyConvention.cs]: One of the reasons I asked for a second model in the test was to deal with potential name collisions.  You have avoided that by creating models with different order by attribute names but I'm still worried this code can't handle the case where two attributes of the same name (from two different entitles in a model OR from two different models) have different ordering. How does your code account for those cases?

File: src/System.Web.OData/OData/Query/ODataQueryOptions.cs [hunk 9]
```diff
@@ -31,13 +31,15 @@ public class ODataQueryOptions
     {
         private static readonly MethodInfo _limitResultsGenericMethod = typeof(ODataQueryOptions).GetMethod("LimitResults");
 
+        private static Dictionary<string, int> _columnOrder = new Dictionary<string, int>();
+
         private ETag _etagIfMatch;
 
         private bool _etagIfMatchChecked;
-
+
```
Comment (robward-ms) [inline src/System.Web.OData/OData/Query/ODataQueryOptions.cs]: avoid trailing whitespace please.

File: src/System.Web.OData/OData/Query/ODataQueryOptions.cs [hunk 10]
```diff
@@ -31,13 +31,15 @@ public class ODataQueryOptions
...
         private ETag _etagIfMatch;
 
         private bool _etagIfMatchChecked;
-
+        
         private ETag _etagIfNoneMatch;
 
-        private bool _etagIfNoneMatchChecked;
+        private bool _etagIfNoneMatchChecked;
```
Comment (robward-ms) [inline src/System.Web.OData/OData/Query/ODataQueryOptions.cs]: avoid trailing whitespace please.

File: src/System.Web.OData/OData/Query/ODataQueryOptions.cs [hunk 11]
```diff
@@ -83,7 +85,22 @@ public ODataQueryOptions(ODataQueryContext context, HttpRequestMessage request)
...
             BuildQueryOptions(queryParameters);
 
-            Validator = ODataQueryValidator.GetODataQueryValidator(context);
+            Validator = ODataQueryValidator.GetODataQueryValidator(context);           
+        }
+
+        /// <summary>
+        /// Gets Column order.
+        /// </summary>
+        public static Dictionary<string, int> ColumnOrder
```
Comment (robward-ms) [inline src/System.Web.OData/OData/Query/ODataQueryOptions.cs]: I get that static is very convenient here but it would be much better if this were stored as a member somewhere that was model or even entity specific.

**Comment by ConstantinOr** [comment]
@robward-ms I was corrected solution. Please help me understand what wrong now.

**Comment by robward-ms** [comment]
@ConstantinOr  -  2098 test failures. Are you using build.cmd to verify your changes before submitting them?

**Comment by ConstantinOr** [comment]
Hmm.  Yes. My test passed, open api passed.

File: src/System.Web.OData/Settings.StyleCop [hunk 12]
```diff
@@ -1,4 +1,4 @@
-<StyleCopSettings Version="105">
+﻿<StyleCopSettings Version="105">
```
Comment (robward-ms) [inline src/System.Web.OData/Settings.StyleCop]: Please revert this change: Settings.StyleCop should not require changes.

File: src/System.Web.OData/System.Web.OData.csproj [hunk 13]
```diff
@@ -540,6 +540,7 @@
   </ItemGroup>
   <ItemGroup>
     <None Include="packages.config" />
+    <None Include="Settings.StyleCop" />
```
Comment (robward-ms) [inline src/System.Web.OData/System.Web.OData.csproj]: Please revert this change: Settings.StyleCop does not need to be part of the project.

**Comment by ConstantinOr** [comment]
@robward-ms I fixed code, per your requests. Please check. Thanks for your help.

File: test/E2ETest/WebStack.QA.Test.OData/ODataOrderByTest/OrderByDataModel.cs [hunk 14]
```diff
@@ -0,0 +1,44 @@
+﻿using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace WebStack.QA.Test.OData.ODataOrderByTest
+{
+    public class Item
+    {
+        [Key]
+        [Column(Order = 2)]
```
Comment (robward-ms) [inline test/E2ETest/WebStack.QA.Test.OData/ODataOrderByTest/OrderByDataModel.cs]: What is the behavior if [Key] but not [Column]? I suspect that key gets ordered first, which may not be the correct behavior?

Comment (ConstantinOr) [reply test/E2ETest/WebStack.QA.Test.OData/ODataOrderByTest/OrderByDataModel.cs]: @robward-ms  Files Updated. I was took logic from the bug. At this moment default "order by", by property name.
If need another logic please describe.

Comment (ConstantinOr) [reply test/E2ETest/WebStack.QA.Test.OData/ODataOrderByTest/OrderByDataModel.cs]: @robward-ms Local build successful.

File: src/System.Web.OData/OData/Query/ODataQueryOptions.cs [hunk 15]
```diff
@@ -31,10 +32,12 @@ public class ODataQueryOptions
     {
         private static readonly MethodInfo _limitResultsGenericMethod = typeof(ODataQueryOptions).GetMethod("LimitResults");
 
+        private static Dictionary<string, int> _columnOrder = new Dictionary<string, int>();
```
Comment (robward-ms) [inline src/System.Web.OData/OData/Query/ODataQueryOptions.cs]: No longer used.

File: src/System.Web.OData/OData/Query/ODataQueryOptions.cs [hunk 16]
```diff
@@ -31,10 +32,12 @@ public class ODataQueryOptions
...
+
         private ETag _etagIfMatch;
 
         private bool _etagIfMatchChecked;
-
+
```
Comment (robward-ms) [inline src/System.Web.OData/OData/Query/ODataQueryOptions.cs]: Please don't add trailing white space.

File: src/System.Web.OData/System.Web.OData.csproj [hunk 17]
```diff
@@ -539,7 +539,7 @@
     <CodeAnalysisDictionary Include="..\CodeAnalysisDictionary.xml" />
   </ItemGroup>
   <ItemGroup>
-    <None Include="packages.config" />
+    <None Include="packages.config" />
```
Comment (robward-ms) [inline src/System.Web.OData/System.Web.OData.csproj]: Please don't add trailing white space.

File: src/System.Web.OData/OData/Query/ODataQueryOptions.cs [hunk 18]
```diff
@@ -518,23 +523,23 @@ private static IEnumerable<IEdmStructuralProperty> GetAvailableOrderByProperties
...
+            var properties =
+                entityType.Key().Any()
+                    ? entityType.Key()
+                    : entityType
+                        .StructuralProperties()
+                        .Where(property => property.Type.IsPrimitive() && !property.Type.IsStream());
+
+            return properties.OrderBy(o =>
+            {
+               var value = o.DeclaringType as PrimitivePropertyConfiguration;
```
Comment (AlanWong-MS) [inline src/System.Web.OData/OData/Query/ODataQueryOptions.cs]: nit: the indentation of this line is different from line 541.

File: test/E2ETest/WebStack.QA.Test.OData/ODataOrderByTest/OrderByEdmModel.cs [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+                : base(ConnectionString)
+            {
+            }
+
+            public IDbSet<Item> Items { get; set; }
+            public IDbSet<Item2> Items2 { get; set; }
+
+        }
+
+
```
Comment (AlanWong-MS) [inline test/E2ETest/WebStack.QA.Test.OData/ODataOrderByTest/OrderByEdmModel.cs]: Please remove extra line.

**Comment by robward-ms** [comment]
@ConstantinOr  - Looks like we're having some build machine issues on our side, I ran your build on 3 machines and got a passing result on 2 of them.

**Comment by ConstantinOr** [comment]
@robward-ms Many thanks for your help and investigation. I was fixed issues.

**Comment by ConstantinOr** [comment]
@robward-ms Again i have problem with build. In local mode works without problem.

**Comment by robward-ms** [comment]
@ConstantinOr  - the database name changed made a difference in the lab, your change now passes on all build machines.
